hilti/src/rt/library.cc

```diff
@@ -1,14 +1,51 @@
 // Copyright (c) 2020 by the Zeek Project. See LICENSE for details.
 
 #include <dlfcn.h>
+#include <nlohmann/json.hpp>
 
+#include <hilti/rt/autogen/version.h>
 #include <hilti/rt/exception.h>
 #include <hilti/rt/library.h>
 #include <hilti/rt/logging.h>
 
 using namespace hilti::rt;
 
-hilti::rt::Library::Library(const std::filesystem::path& path) {
+std::string hilti::rt::library::Version::toJSON() const {
+    return nlohmann::json{{"magic", magic},
+                          {"hilti_version", hilti_version},
+                          {"created", created},
+                          {"debug", debug},
+                          {"optimize", optimize}}
+        .dump();
+}
+
+hilti::rt::Result<Nothing> hilti::rt::library::Version::fromJSON(const std::string& json) {
+    try {
+        auto j = nlohmann::json::parse(json);
+        j.at("magic").get_to(magic);
+        j.at("hilti_version").get_to(hilti_version);
+        j.at("created").get_to(created);
+        j.at("debug").get_to(debug);
+        j.at("optimize").get_to(optimize);
+        return Nothing();
+    } catch ( const nlohmann::json::exception& e ) {
+        return result::Error(e.what());
+    }
+}
+
+void hilti::rt::library::Version::checkCompatibility() const {
+    if ( hilti_version != PROJECT_VERSION_NUMBER )
+        warning(fmt("module %s was compiled with HILTI version %d, but using HILTI version %d", path.filename(),
+                    hilti_version, PROJECT_VERSION_NUMBER));
+
+    if ( hilti::rt::isDebugVersion() && optimize )
```

Comment:
This is essentially still the tech debt of us conflating debug and/or optimizations in `debug`. Ideally this would just check if we are unoptimized right now and we should at least capture that in a comment.

Do you think we'd want a similar warning for the inverse (running an unoptimized library while in an optimized run)?

Reply:
Yeah, that debug vs opt is still on my mind, and I was tempted to address it in the same PR, but then didn't want to conflate things.

I opted against a warning for the inverse as it seems a more common use case for users. They'll have the release build of the toolchain by default, and may just be debugging their parsers.